Source/JavaScriptCore/builtins/JSIteratorConstructor.js

```diff
@@ -71,3 +71,44 @@ function from(value)
     // 6. Return wrapper.
     return @wrapForValidIteratorCreate(iterator, iteratorNextMethod);
 }
+
+// https://tc39.es/proposal-iterator-sequencing/#sec-iterator.concat
+function concat(/* ...items */)
+{
+    "use strict";
+    var len = @argumentCount();
+
+    var iterables = [];
+    for (var i = 0; i < @argumentCount(); ++i) {
+        var iterable = arguments[i];
+        if (!@isObject(iterable))
+            @throwTypeError("Iterator.concat expects all arguments to be objects");
+
+        var openMethod = iterable.@@iterator;
+        if (@isUndefinedOrNull(openMethod))
+            @throwTypeError("Iterator.concat expects all arguments to be iterable");
+
+        @putByValDirect(iterables, i, {openMethod, iterable});
+    }
+
+    var generator = (function*() {
+        for (var iterable of iterables) {
+            var iterator = iterable.openMethod.@call(iterable.iterable);
+            if (!@isObject(iterator))
+                @throwTypeError("Iterator.concat expects all arguments to be iterable");
+
+            for (;;) {
+                var result = @iteratorGenericNext(iterator.next, iterator);
+                if (result.done)
+                    break;
+
+                var value = result.value;
+                @ifAbruptCloseIterator(iterator, (
+                    yield value
+                ));
+            }
+        }
+    })();
+
+    return @iteratorHelperCreate(generator, null);
```

Comment:
FYI: According to the spec, if the `IteratorHelperPrototype` object doesn't have an `[[UnderlyingIterator]]`, the return method should throw an error, but I wasn't sure if this was the intended behavior, so I created an issue in the spec: https://github.com/tc39/proposal-iterator-sequencing/issues/14

Reply:
oh neat yeah i was gonna ask a similar question but then saw that someone (which i now know was you) already did.  thanks! :)

Reply:
IteratorHelper's `return` will throw a TypeError if `[[UnderlyingIterator]]` is `null`. Please add a `null` check there and a test.